packages/SwingSet/src/kernel/vatManager/snapStore.js

```diff
@@ -0,0 +1,127 @@
+ *   createWriteStream: typeof import('fs').createWriteStream,
+ *   resolve: typeof import('path').resolve,
+ *   rename: typeof import('fs').promises.rename,
+ *   unlink: typeof import('fs').promises.unlink,
+ * }} io
+ */
+export function makeSnapstore(
+  root,
+  {
+    tmpName,
+    existsSync,
+    createReadStream,
+    createWriteStream,
+    resolve,
+    rename,
+    unlink,
+  },
+) {
+  /** @type {(opts: unknown) => Promise<string>} */
+  const ptmpName = promisify(tmpName);
+  const tmpOpts = { tmpdir: root, template: 'tmp-XXXXXX.xss' };
+  /**
+   * @param { (name: string) => Promise<T> } thunk
+   * @returns { Promise<T> }
+   * @template T
+   */
+  async function withTempName(thunk) {
+    const name = await ptmpName(tmpOpts);
+    const result = await thunk(name);
+    try {
+      await unlink(name);
+    } catch (ignore) {
+      // ignore
+    }
+    return result;
+  }
+
+  /**
+   * @param {string} dest
+   * @param { (name: string) => Promise<T> } thunk
+   * @returns { Promise<T> }
+   * @template T
+   */
+  async function atomicWrite(dest, thunk) {
+    const tmp = await ptmpName(tmpOpts);
+    const result = await thunk(tmp);
+    await rename(tmp, resolve(root, dest));
+    try {
+      await unlink(tmp);
+    } catch (ignore) {
+      // ignore
+    }
+    return result;
+  }
+
+  /** @type {(input: string, f: NodeJS.ReadWriteStream, output: string) => Promise<void>} */
+  async function filter(input, f, output) {
+    const source = createReadStream(input);
+    const destination = createWriteStream(output);
+    await pipe(source, f, destination);
+  }
+
+  /** @type {(filename: string) => Promise<string>} */
+  async function fileHash(filename) {
+    const h = createHash('sha256');
+    const p = createReadStream(filename).pipe(h);
+    await new Promise(cb => p.end(cb));
+    return h.digest('hex');
+  }
+
+  /**
+   * @param {(fn: string) => Promise<void>} saveRaw
+   * @returns { Promise<string> } sha256 hash of (uncompressed) snapshot
+   */
+  async function save(saveRaw) {
+    return withTempName(async snapFile => {
+      await saveRaw(snapFile);
+      const h = await fileHash(snapFile);
+      if (existsSync(`${h}.gz`)) return h;
+      await atomicWrite(`${h}.gz`, gztmp =>
+        filter(snapFile, createGzip(), gztmp),
+      );
+      return h;
+    });
+  }
+
+  /**
+   * @param {string} hash
+   * @param {(fn: string) => Promise<T>} loadRaw
+   * @template T
+   */
+  async function load(hash, loadRaw) {
+    return withTempName(async raw => {
+      await filter(resolve(root, `${hash}.gz`), createGunzip(), raw);
+      const actual = await fileHash(raw);
+      assert(actual === hash, d`actual hash ${actual} !== expected ${hash}`);
+      // be sure to await loadRaw before exiting withTempName
+      const result = await loadRaw(raw);
+      return result;
+    });
+  }
+
+  return freeze({ load, save });
+}
+
+export function defaultSnapstorePath({ env = process.env }) {
```

Comment:
As above, let's remove the knowledge of suitable directories from this code and have the caller make that decision. In both ag-solo and cosmic-swingset, we're going to be putting these snapshots under the same directory that holds the rest of our kernel state, which will be something specific to the host app, rather than `~/.cache` anyways.